### PR TITLE
Add login, logout for scope based operations (#917)

### PR DESCRIPTION
**Pull request by hishnash**
See issue: #917

added:
* `login(scope, user)`
* `logout(scope)`

updated:
* `get_user(scope)`

this removed the `_cached_user` key from `scope` since if you are logging in and out on another level in the stack we cant assume that modifications to the scope on that level will be acceptable to the scope in other levels so I think it best to always fall back to checking the session.

---

TODO:
- [x] Tests
- [-] CSRF_COOKIE/Token middleware will do this as part of a separate pull request
- [x] Add docs

### CODE REVIEW

**Comment by hishnash** [comment]
@andrewgodwin let me know if there are any other changes if not I will do a rebase to compress all the commits.

**Comment by andrewgodwin** [comment]
There's no need to rebase, github has squash merge built in. I'll take a look later.

**Comment by andrewgodwin** [comment]
Hmm, looking over this once more I have noticed the `async_login` and `async_logout` functions - I'd rather just let the user wrap the ones provided as appropriate, so remove them, but mention in the docs that they need to wrap them if calling them from an async context.

**Comment by hishnash** [comment]
@andrewgodwin done but im wanding if I should add `database_sync_to_async` to the call of `get_user` in the `AuthMiddleware` here: https://github.com/hishnash/channels/blob/7cf0ce8f38634b286c7ec863e6a49c50dc6832b4/channels/auth.py#L136

Since it depends on the user's Authentication Backend as it if this is needed.

**Comment by andrewgodwin** [comment]
Hmm, you can't use lazy objects because you might traverse a `database_sync_to_async` call, so everything would have to be done in the function. I am almost tempted to say these should be async functions entirely, to continue with Channels being outwardly async (even if we switch to sync code inside).

**Comment by hishnash** [comment]
so to be clear the AuthMidleware's `__call__` (the same as the __init__ of a Consumer) is called inside a `SyncToAsync` so it would be ok to have an `AsyncToSync`.

**Comment by hishnash** [comment]
just wrap in `database_sync_to_async` were it is likly that some subclass of User/AuthBackend might hit the db:
```python
async def login(scope, user, backend=None):
    """
    Persist a user id and a backend in the request.
    This way a user doesn't have to re-authenticate on every request.
    Note that data set during the anonymous session is retained when the user logs in.
    """
    if "session" not in scope:
        raise ValueError("Cannot find session in scope. You should wrap your consumer in SessionMiddleware.")
    session = scope["session"]
    session_auth_hash = ""
    if user is None:
        user = scope.get("user", None)
    if user is None:
        raise ValueError("User must be passed as an argument or must be present in the scope.")
    if hasattr(user, "get_session_auth_hash"):
        session_auth_hash = await database_sync_to_async(user.get_session_auth_hash)()
    if SESSION_KEY in session:
        if await database_sync_to_async(_get_user_session_key)(session) != user.pk or (
                session_auth_hash and not
                constant_time_compare(session.get(HASH_SESSION_KEY, ""), session_auth_hash)):
            # To avoid reusing another user's session, create a new, empty
            # session if the existing session corresponds to a different
            # authenticated user.
            await database_sync_to_async(session.flush)()
    else:
        await database_sync_to_async(session.cycle_key)()
    try:
        backend = backend or user.backend
    except AttributeError:
        backends = await database_sync_to_async(_get_backends)(return_tuples=True)
        if len(backends) == 1:
            _, backend = backends[0]
        else:
            raise ValueError(
                "You have multiple authentication backends configured and therefore must provide the `backend` "
                "argument or set the `backend` attribute on the user."
            )
    session[SESSION_KEY] = await database_sync_to_async(user._meta.pk.value_to_string)(user)
    session[BACKEND_SESSION_KEY] = backend
    session[HASH_SESSION_KEY] = session_auth_hash
    scope["user"] = user
    # note this does not reset the CSRF_COOKIE/Token
    await database_sync_to_async(user_logged_in.send)(sender=user.__class__, request=None, user=user)
```

or wrap the entier thing
```python
async def login(scope, user, backend=None):
    @database_sync_to_async
    def _login(.....):
          ....
    return await _login(......)
```

**Comment by hishnash** [comment]
@andrewgodwin I switched to `async` methods so you can see how it feels (can revert back if you feel you prefer the sync in the end)

**Comment by andrewgodwin** [comment]
Hmm, I really don't know which one is better at this point. Give me the weekend to think about it.

File: channels/auth.py [hunk 1]
```diff
@@ -28,7 +137,7 @@ def __call__(self, scope):
             raise ValueError("AuthMiddleware cannot find session in scope. SessionMiddleware must be above it.")
         # Add it to the scope if it's not there already
         if "user" not in scope:
-            scope["user"] = SimpleLazyObject(lambda: get_user(scope))
+            scope["user"] = async_to_sync(get_user)(scope)
```
Comment (andrewgodwin) [inline channels/auth.py]: You can't do this here, as you'd be doing a blocking operation in the ASGI app's first callable (which can't block). I'm surprised this didn't raise an error during testing... still, you should keep the lambda for now (that means you can still accidentally call it during an async consumer, but I'd like to fix that separately)

File: docs/topics/sessions.rst [hunk 2]
```diff
@@ -75,3 +75,52 @@ connections or HTTP views won't be able to see the changes.
     If you are in a long-polling HTTP consumer, you might want to save changes
     to the session before you send a response. If you want to do this,
     call ``scope["session"].save()``.
+
+
+How to log a user in/out
+------------------------
+
+Within your consumer you can call ``login(scope, user, backend=None)`` to login a user.
```
Comment (andrewgodwin) [inline docs/topics/sessions.rst]: call -> await
"to login a user" -> "to log a user in"

File: docs/topics/sessions.rst [hunk 3]
```diff
@@ -75,3 +75,52 @@ connections or HTTP views won't be able to see the changes.
...
+This requires that your scope has a ``session`` object the best way to do this is to
```
Comment (andrewgodwin) [inline docs/topics/sessions.rst]: "object the best way" -> "object; the best way"

File: docs/topics/sessions.rst [hunk 4]
```diff
@@ -75,3 +75,52 @@ connections or HTTP views won't be able to see the changes.
...
+ensure your consumer is wrapped in a ``SessionMiddlewareStack`` or a ``AuthMiddlewareStack``.
+
+You can logout a user with the ``logout(scope)`` function.
```
Comment (andrewgodwin) [inline docs/topics/sessions.rst]: "function" -> "async function"

File: docs/topics/sessions.rst [hunk 5]
```diff
@@ -75,3 +75,52 @@ connections or HTTP views won't be able to see the changes.
...
+
+If you are in a WebSocket consumer, or logging-in after the first response has been sent in a http consumer,
+the session is populated **but will never be saved automatically** - you must call ``scope["session"].save()``
+after login in your consumer code::
+
```
Comment (andrewgodwin) [inline docs/topics/sessions.rst]: Remove extra newline

File: docs/topics/sessions.rst [hunk 6]
```diff
@@ -75,3 +75,52 @@ connections or HTTP views won't be able to see the changes.
...
+    class ChatConsumer(AsyncWebsocketConsumer):
+
+        ...
+
+        async def receive(self, text_data):
+            ...
+            # login the user to this session.
+            await login(self.scope, user)
+            # save the session (if the session backend does not access the db you an use `sync_to_async`)
```
Comment (andrewgodwin) [inline docs/topics/sessions.rst]: "you an use" -> "you can use"

File: docs/topics/sessions.rst [hunk 7]
```diff
@@ -75,3 +75,52 @@ connections or HTTP views won't be able to see the changes.
...
+            await database_sync_to_async(self.scope["session"].save)()
+
+When calling ``login(scope, user)``, ``logout(scope)`` or ``get_user(scope)`` from an sync function you will need to
+wrap them in ``async_to_sync``::
+
```
Comment (andrewgodwin) [inline docs/topics/sessions.rst]: Remove extra newline

File: docs/topics/sessions.rst [hunk 8]
```diff
@@ -75,3 +75,52 @@ connections or HTTP views won't be able to see the changes.
...
+        def receive(self, text_data):
+            ...
+            async_to_sync(login)(self.scope, user)
+            self.scope["session"].save()
+
+.. note::
+
+    If you are using a long running consumer, websocket or long-polling HTTP it is possible
+    the user has been logged out of their session elsewere.
```
Comment (andrewgodwin) [inline docs/topics/sessions.rst]: elsewere -> elsewhere

File: docs/topics/sessions.rst [hunk 9]
```diff
@@ -75,3 +75,52 @@ connections or HTTP views won't be able to see the changes.
...
+    Use ``get_user(scope)`` to be sure that the user is still logged in.
```
Comment (andrewgodwin) [inline docs/topics/sessions.rst]: "Use" -> "You can periodically use"

File: tests/security/test_auth.py [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+    assert await get_user(scope) == user
+
+
+@pytest.mark.django_db(transaction=True)
+@pytest.mark.asyncio
+async def test_login_no_session_in_scope():
+
+    with pytest.raises(
+            ValueError,
+            match="Cannot find session in scope. You should wrap your consumer in SessionMiddleware."):
```
Comment (andrewgodwin) [inline tests/security/test_auth.py]: Move closing bracket to next line

File: tests/security/test_auth.py [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+    assert BACKEND_SESSION_KEY in session
+    assert HASH_SESSION_KEY in session
+
+    assert isinstance(await get_user(scope), await database_sync_to_async(get_user_model)())
+    assert await get_user(scope) == user
+
+
+@pytest.mark.django_db(transaction=True)
+@pytest.mark.asyncio
+async def test_login_no_session_in_scope():
```
Comment (andrewgodwin) [inline tests/security/test_auth.py]: Docstrings on each test please saying what it is testing as an English sentence

File: tests/security/test_auth.py [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+    BACKEND_SESSION_KEY, HASH_SESSION_KEY, SESSION_KEY, get_user_model, user_logged_in, user_logged_out,
+)
+from django.contrib.auth.models import AnonymousUser
+
+from asgiref.sync import sync_to_async
+from channels.auth import get_user, login, logout
+from channels.db import database_sync_to_async
+
+
+class catch_signal():
```
Comment (andrewgodwin) [inline tests/security/test_auth.py]: Docstring saying what this does please

**Comment by hishnash** [comment]
@andrewgodwin let me know if there are any other changes?

**Comment by hishnash** [comment]
@andrewgodwin ... should be good now 🤞

**Comment by andrewgodwin** [comment]
Great, thanks!

**Pull request merged**